[CUBRIDMAN-189] SQL Rewriter : Predicate push (#452)
CUBRID/cubrid-manual: ko/sql/tuning.rst

                8           60          600  'Left- 706'                     6
                9           60          600  'Left- 806'                     6
               10           60          600  'Left- 906'                     6

> .. _pred_push

Reviewer: ```suggestion

.. _pred_push
```

조건절 푸시 최적화
-----------------------

> 조건절 푸시(Predicate Push)는 뷰를 조회하기 전에 먼저 조건절을 적용한다. 

Reviewer: How about using only `Predicate Push` instead of `조건절 푸시` ?

Author: I modified `조건절 푸시` to `Predicate Push`

Reviewer: How about `**Predicate Push**` instead of `Predicate Push` ?

Author: I modified `Predicate Push` to `**Predicate Push**`.

이를 통해 조건에 만족하는 더 적은 양의 데이터만 조회되도록 하여 전체 처리량을 줄일 수 있다.

예를 들어 **GROUP BY**\절을 포함한 아래 뷰를 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면

> **GROUP BY**\  해야할 데이터양을 줄일 수 있다. 

Reviewer: ```suggestion
**GROUP BY** 해야 할 데이터양을 줄일 수 있다. 
```

.. code-block:: sql

    SELECT b.deptno, b.dname, a.avg_sal

    FROM (SELECT deptno, avg(sal) avg_sal FROM emp GROUP BY deptno) a

        , dept b

    WHERE a.deptno = b.deptno

>     AND a.deptno = 30;

Reviewer: `emp` and `dept` are not cubrid example tables.
To keep it, the `CREATE TABLE` statement should be added.

Author: I modified it to cubrid demodb example.

>     AND a.deptno = 30;

Reviewer: I think the empty line is unnecessary.
Was it intended?

Author: Nope, I removed empty line

위 쿼리에 정의한 뷰 내부에는 조건절이 없다. 만약 쿼리 변환이 작동하지 않는다면,

> *emp*\  테이블을 Full Scan 하고서 **GROUP BY**\ 이후에 *deptno = 30*\ 조건을

Reviewer: ```suggestion
*emp* 테이블을 Full Scan 하고서 **GROUP BY** 이후에 *deptno = 30* 조건을
```

필터링했을 것이다. 

하지만 조건절 푸시를 통해서 다음과 같이 쿼리가 변환된다면, 더 적은 양의
데이터만 조회되도록 최적화할 수 있다.

.. code-block:: sql

    SELECT b.deptno, b.dname, a.avg_sal

    FROM (SELECT deptno, avg(sal) avg_sal FROM emp GROUP BY deptno WHERE deptno = 30) a

        , dept b

    WHERE a.deptno = b.deptno

    AND a.deptno = 30;

다음의 경우엔 조건절 푸시 최적화가 수행되지 않는다.

    #. **CONNECT BY**\를 포함한 경우

    #. 집계함수나 분석함수를 사용하는 경우

>     #. **ROWNUM, LIMIT**\ 또는 **GROUPBY_NUM (), INST_NUM (), ORDERBY_NUM ()**\ 을 사용하는 경우

Reviewer: ```suggestion
    #. **ROWNUM, LIMIT** 또는 **GROUPBY_NUM (), INST_NUM (), ORDERBY_NUM ()**\ 을 사용하는 경우
```

    #. **Correlated Subquery**\ 를 사용하여 작성된 경우

>     #. OUTER JOIN의 **ON**\절에 조건절이 작성된 경우

Reviewer: ```suggestion
    #. **OUTER JOIN**\의 **ON**\절에 조건절이 작성된 경우
```
Consistency of expression with #450 was not maintained.

Reviewer: Other things need to change too.

Author: I changed other OUTER JOIN

    #. OUTER JOIN을 수행할 때 푸시될 조건절이나 뷰 내부의 조건절 푸시 대상에 NULL 변환 함수를 사용할 경우

    #. 조건절에 부질의가 사용된 경우

    #. 뷰가 메소드를 포함한 경우

    #. OUTER JOIN을 수행할 때 푸시될 조건절이나 뷰 내부의 조건절 푸시 대상에 부질의를 사용하는 경우

>     #. OUTER JOIN을 수행할 때 푸시될 조건절이나 뷰 내부의 조건절 푸시 대상에 **RANDOM (), DRANDOM (), SYS_GUID ()**를 사용하는 경우

Reviewer: ```suggestion
    #. OUTER JOIN을 수행할 때 푸시될 조건절이나 뷰 내부의 조건절 푸시 대상에 **RANDOM (), DRANDOM (), SYS_GUID ()**\를 사용하는 경우
```

>     #. OUTER JOIN을 수행할 때 푸시될 조건절이나 뷰 내부의 조건절 푸시 대상에 **RANDOM (), DRANDOM (), SYS_GUID ()**를 사용하는 경우

Reviewer: "OUTER JOIN을 수행할 때 푸시될 조건절이나 뷰 내부의 조건절 푸시 대상에" is used repeatedly.
What about collecting and listing things for this category?

Author: I have placed three items regarding OUTER JOIN at bottom.

Reviewer: This does not mean that the sorting in the list should be changed.
This means reducing the repetition of sentences.

Author: I reduced `OUTER JOIN` repetition

Reviewer: What I thought is as follows:
```
    #. **OUTER JOIN**\을 수행할 때 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 다음을 사용하는 경우:
        * **ON**\절에 조건절이 작성된 경우
        * **NULL** 변환 함수
        * 부질의를 사용하는 경우
        * **RANDOM (), DRANDOM (), SYS_GUID ()**\를 사용하는 경우
```
The above content is not smooth, so please refer to it only.

다음은 질의가 OUTER JOIN을 수행할 때 **ON**\절의 조건에 푸시될 조건절이 있는 예시이다.

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT * FROM athlete WHERE nation_code = 'KOR') a
        LEFT OUTER JOIN record r 
        ON a.code = r.athlete_code;

> 이 경우, a.nation_code = 'KOR'는 LEFT OUTER JOIN 수행 시 ON 절에 있는데, 이러한 형태로 ON 절의 조건절은 조건절 푸시 대상이 아니다.

Reviewer: I think the example is wrong.
There is no predicate `a.nation_code = 'KOR'` .

Author: I modified it to right example.

다음 질의는 OUTER JOIN을 수행할 때 푸시될 조건절이나 뷰 내부의 조건절 푸시 대상에 NULL 변환 함수를 사용하는 예시이다.

.. code-block:: sql

        SELECT a.name, r.score 
        FROM athlete a
        LEFT OUTER JOIN (SELECT * FROM record) r 
        ON a.code = r.athlete_code
        WHERE NVL(r.score, '0') = '0';

OUTER JOIN을 수행할 때 푸시될 조건절이나 뷰 내부의 조건절 푸시 대상에 NULL 변환 함수를 사용한 경우 조건절 푸시 대상이 아니다.
NULL 변환 함수에는 **COALESCE (), NVL (), NVL2 (), DECODE (), IF (), IFNULL (), CONCAT_WS ()** 가 포함된다.

> 또한 **IS NULL, CASE** 문 또한 NULL 변환 함수로 취급된다.

Reviewer: How about using a different word than `함수` ?
`**IS NULL, CASE**` is not a function.

Author: I modified it to `또한 **IS NULL, CASE** 문 또한 Predicate Push 대상이 아니다.`

                8           60          600  'Left- 706'                     6
                9           60          600  'Left- 806'                     6
               10           60          600  'Left- 906'                     6
.. _pred_push

조건절 푸시 최적화
-----------------------

> 조건절 푸시(Predicate Push)는 뷰를 조인하기 전에 먼저 조건절을 적용한다. 

Reviewer 2: ```suggestion
조건절 푸시(Predicate Push)는 뷰를 조회하기 전에 먼저 조건절을 적용한다. 
```

> 이를 통해 조건에 만족하는 더 적은 양의 데이터만 조인되도록 하여 전체 처리량을 줄일 수 있다.

Reviewer 2: ```suggestion
이를 통해 조건에 만족하는 더 적은 양의 데이터만 조회되도록 하여 전체 처리량을 줄일 수 있다.
```

예를 들어 **GROUP BY**\절을 포함한 아래 뷰를 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면
**GROUP BY**\  해야할 데이터양을 줄일 수 있다. 

.. code-block:: sql

    SELECT b.deptno, b.dname, a.avg_sal

    FROM (SELECT deptno, avg(sal) avg_sal FROM emp GROUP BY deptno) a

        , dept b

    WHERE a.deptno = b.deptno

    AND a.deptno = 30;

위 쿼리에 정의한 뷰 내부에는 조건절이 없다. 만약 쿼리 변환이 작동하지 않는다면,
*emp*\  테이블을 Full Scan 하고서 **GROUP BY**\ 이후에 *deptno = 30*\ 조건을
필터링했을 것이다. 

하지만 조건절 푸시를 통해서 다음과 같이 쿼리가 변환된다면, 더 적은 양의

> 데이터만 조인되도록 최적화할 수 있다.

Reviewer 2: ```suggestion
데이터만 조회되도록 최적화할 수 있다.
```

.. code-block:: sql

    SELECT b.deptno, b.dname, a.avg_sal

    FROM (SELECT deptno, avg(sal) avg_sal FROM emp GROUP BY deptno WHERE deptno = 30) a

        , dept b

    WHERE a.deptno = b.deptno

    AND a.deptno = 30;

다음의 경우엔 조건절 푸시 최적화가 수행되지 않는다.

    #. **CONNECT BY**\를 포함한 경우

    #. 집계함수나 분석함수를 사용하는 경우

    #. **ROWNUM, LIMIT**\ 또는 **GROUPBY_NUM (), INST_NUM (), ORDERBY_NUM ()**\ 을 사용하는 경우

    #. **Correlated Subquery**\ 를 사용하여 작성된 경우

>     #. **JOIN ~ ON**\ 절을 이용하여 OUTER JOIN을 수행할 때 **ON**\절의 조건에 푸시될 조건절이 and로 결합되어 있는 경우

Reviewer 2: ```suggestion
    #. OUTER JOIN의 **ON**\절에 조건절이 작성된 경우
```

    #. **WHERE**\ 절에서 **NVL()**\ 함수의 반환값을 상수와 비교하는 경우 

>     #. 조건절에 쿼리가 사용된 경우

Reviewer 2: ```suggestion
    #. 조건절에 부질의가 사용된 경우
```

    #. 뷰가 메소드를 포함한 경우

    #. 뷰의 **SELECT**\ 리스트에 **SELECT**\ 문이 포함된 경우

    #. 뷰가 **RANDOM (), DRANDOM (), SYS_GUID ()**\를 포함한 경우

> 다음은 질의가 **JOIN ~ ON**\ 절을 이용하여 OUTER JOIN을 수행할 때 **ON**\절의 조건에 푸시될 조건절이 and로 결합되어 있는 예시이다.

Reviewer 2: ```suggestion
다음은 질의가 OUTER JOIN을 수행할 때 **ON**\절의 조건에 푸시될 조건절이 있는 예시이다.
```

.. code-block:: sql

        SELECT a.name, r.score 
        FROM athlete a
        LEFT OUTER JOIN record r 

>         ON a.code = r.athlete_code AND a.nation_code = 'KOR'

Reviewer 2: inline view is needed.

Author: I added inline view.

        WHERE r.medal = 'G';

> 이 경우, a.nation_code = 'KOR'는 LEFT OUTER JOIN 수행 시 ON 절에 있는데, 이러한 형태로 ON 절에 조건절이 포함되면 조건절 푸시 최적화가 일어나지 않을 수 있다.

Reviewer 2: ```suggestion
이 경우, a.nation_code = 'KOR'는 LEFT OUTER JOIN 수행 시 ON 절에 있는데, 이러한 형태로 ON 절에 조건절은 조건절 푸시 대상이 아니다.
```

> 다음 질의는 OUTER JOIN 시 WHERE 절에서 NVL() 함수의 반환값을 상수와 비교하는 예시이다.

Reviewer 2: nullable function. check pt_has_nullable_term ()

Author: I added nullable function example.

*athlete* 테이블을 Full Scan 하고서 조인을 수행한 이후에 *a.nation_code = 'KOR'* 조건을
필터링했을 것이다. 

하지만 **Predicate Push**\를 통해서 다음과 같이 쿼리가 변환된다면, 더 적은 양의
데이터만 조회되도록 최적화할 수 있다.

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT name, nation_code, count(*) cnt FROM athlete WHERE nation_code = 'KOR' GROUP BY name, nation_code ) a, record r
        WHERE a.code = r.athlete_code;

다음의 경우에는 **Predicate Push**\가 수행되지 않는다.

    #. **CONNECT BY**\를 포함한 경우

    #. 집계함수나 분석함수를 사용하는 경우

    #. **ROWNUM, LIMIT** 또는 **GROUPBY_NUM (), INST_NUM (), ORDERBY_NUM ()**\ 을 사용하는 경우

    #. **Correlated Subquery**\ 를 사용하여 작성된 경우

    #. 조건절에 부질의가 사용된 경우

    #. 뷰가 메소드를 포함한 경우

    #. 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 **RANDOM (), DRANDOM (), SYS_GUID ()**\를 사용하는 경우

    #. **OUTER JOIN**\을 수행할 때 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 다음을 사용하는 경우:
            * **ON**\절에 조건절이 작성된 경우
            * **NULL** 변환 함수

다음은 질의가 **OUTER JOIN**\을 수행할 때 **ON**\절의 조건에 푸시될 조건절이 있는 예시이다.

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT * FROM athlete WHERE gender = 'M') a
            LEFT OUTER JOIN record r ON a.code = r.athlete_code AND a.nation_code = 'KOR';

> 이 경우, *a.nation_code = 'KOR'*\는 LEFT **OUTER JOIN** 수행 시 **ON** 절에 있는데, 이러한 형태로 **ON** 절의 조건절은 **Predicate Push** 대상이 아니다.

Reviewer 3: ```suggestion
이 경우, *a.nation_code = 'KOR'*\는 **LEFT OUTER JOIN** 수행 시 **ON** 절에 있는데, 이러한 형태로 **ON** 절의 조건절은 **Predicate Push** 대상이 아니다.
```

다음 질의는 **OUTER JOIN**\을 수행할 때 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 **NULL** 변환 함수를 사용하는 예시이다.

.. code-block:: sql

        SELECT a.name, r.score 
        FROM athlete a
                LEFT OUTER JOIN (SELECT * FROM record WHERE medal = 'G') r ON a.code = r.athlete_code
        WHERE NVL(r.score, '0') = '0';

**OUTER JOIN**\을 수행할 때 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 **NULL** 변환 함수를 사용한 경우 **Predicate Push** 대상이 아니다.

> NULL 변환 함수에는 **COALESCE (), NVL (), NVL2 (), DECODE (), IF (), IFNULL (), CONCAT_WS ()** 가 포함된다.

Reviewer 3: ```suggestion
**NULL** 변환 함수에는 **COALESCE (), NVL (), NVL2 (), DECODE (), IF (), IFNULL (), CONCAT_WS ()** 가 포함된다.
```

                9           60          600  'Left- 806'                     6
               10           60          600  'Left- 906'                     6

.. _pred-push:

Predicate Push
-----------------------
**Predicate Push**\는 외부의 조건절을 뷰 안으로 밀어 넣는 최적화이다.

이를 통해 조건에 만족하는 더 적은 양의 데이터만 조회되도록 하여 전체 처리량을 줄일 수 있다.

예를 들어 아래 쿼리 처럼 *a.code = r.athlete_code*\의 조인조건으로 조인을 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면
조인해야 할 데이터양을 줄일 수 있다. 

.. code-block:: sql

        SELECT a.name, r.score 

>         FROM (SELECT name, nation_code, count(*) cnt FROM athlete GROUP BY name, nation_code) a, record r

Reviewer 2: ```suggestion
        FROM (SELECT name, nation_code, count(*) cnt 
                     FROM athlete 
                    GROUP BY name, nation_code) a, 
                    record r
```

        WHERE a.code = r.athlete_code
        AND a.nation_code = 'KOR';

위 쿼리에 정의한 뷰 내부에는 조건절이 없다. 만약 쿼리 변환이 작동하지 않는다면,
*athlete* 테이블을 Full Scan 하고서 조인을 수행한 이후에 *a.nation_code = 'KOR'* 조건을
필터링했을 것이다. 

하지만 **Predicate Push**\를 통해서 다음과 같이 쿼리가 변환된다면, 더 적은 양의
데이터만 조회되도록 최적화할 수 있다.

.. code-block:: sql

        SELECT a.name, r.score 

>         FROM (SELECT name, nation_code, count(*) cnt FROM athlete WHERE nation_code = 'KOR' GROUP BY name, nation_code ) a, record r

Reviewer 2: ```suggestion
        FROM (SELECT name, nation_code, count(*) cnt FROM athlete 
        WHERE nation_code = 'KOR' GROUP BY name, nation_code ) a, record r
```

                9           60          600  'Left- 806'                     6
               10           60          600  'Left- 906'                     6

.. _pred-push:

Predicate Push
-----------------------
**Predicate Push**\는 외부의 조건절을 뷰 안으로 밀어 넣는 최적화이다.

이를 통해 조건에 만족하는 더 적은 양의 데이터만 조회되도록 하여 전체 처리량을 줄일 수 있다.

> 예를 들어 아래 쿼리 처럼 *a.code = r.athlete_code*\절을 사용한 조인을 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면

Reviewer 2: ```suggestion
예를 들어 아래 쿼리 처럼 *a.code = r.athlete_code*\의 조인조건으로 조인을 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면
```
'절' is SELECT or FROM ...    'a.code = r.ath...' is predicate (조인조건)

조인해야 할 데이터양을 줄일 수 있다. 

.. code-block:: sql

        SELECT a.name, r.score 

>         FROM (SELECT * FROM athlete) a, record r

Reviewer 2: ```suggestion
        FROM (SELECT name, count(*) cnt FROM athlete GROUP BY name) a, record r
```
Examples from actual use

>         FROM (SELECT name, count(*) cnt FROM athlete GROUP BY name) a, record r

Reviewer 2: ```suggestion
        FROM (SELECT name, nation_code, count(*) cnt FROM athlete GROUP BY name, nation_code) a, record r
```

        WHERE a.code = r.athlete_code
        AND a.nation_code = 'KOR';

위 쿼리에 정의한 뷰 내부에는 조건절이 없다. 만약 쿼리 변환이 작동하지 않는다면,
*athlete* 테이블을 Full Scan 하고서 조인을 수행한 이후에 *a.nation_code = 'KOR'* 조건을
필터링했을 것이다. 

하지만 **Predicate Push**\를 통해서 다음과 같이 쿼리가 변환된다면, 더 적은 양의
데이터만 조회되도록 최적화할 수 있다.

.. code-block:: sql

        SELECT a.name, r.score 

>         FROM (SELECT * FROM athlete WHERE nation_code = 'KOR') a, record r

Reviewer 2: ```suggestion
        FROM (SELECT name, count(*) cnt FROM athlete WHERE nation_code = 'KOR' GROUP BY name ) a, record r
```

>         FROM (SELECT name, count(*) cnt FROM athlete WHERE nation_code = 'KOR' GROUP BY name ) a, record r

Reviewer 2: ```suggestion
        FROM (SELECT name, nation_code, count(*) cnt FROM athlete WHERE nation_code = 'KOR' GROUP BY name, nation_code ) a, record r
```

                9           60          600  'Left- 806'                     6
               10           60          600  'Left- 906'                     6

.. _pred-push:

Predicate Push
-----------------------
**Predicate Push**\는 외부의 조건절을 뷰 안으로 밀어 넣는 최적화이다.

이를 통해 조건에 만족하는 더 적은 양의 데이터만 조회되도록 하여 전체 처리량을 줄일 수 있다.

예를 들어 아래 쿼리 처럼 *a.code = r.athlete_code*\절을 사용한 조인을 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면
조인해야 할 데이터양을 줄일 수 있다. 

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT * FROM athlete) a, record r
        WHERE a.code = r.athlete_code
        AND a.nation_code = 'KOR';

위 쿼리에 정의한 뷰 내부에는 조건절이 없다. 만약 쿼리 변환이 작동하지 않는다면,
*athlete* 테이블을 Full Scan 하고서 조인을 수행한 이후에 *a.nation_code = 'KOR'* 조건을
필터링했을 것이다. 

하지만 **Predicate Push**\를 통해서 다음과 같이 쿼리가 변환된다면, 더 적은 양의
데이터만 조회되도록 최적화할 수 있다.

.. code-block:: sql

        SELECT a.name, r.score 

>         FROM (SELECT * FROM athlete WHERE a.nation_code = 'KOR') a, record r

Reviewer: ```suggestion
        FROM (SELECT * FROM athlete WHERE nation_code = 'KOR') a, record r
```

        WHERE a.code = r.athlete_code
        AND a.nation_code = 'KOR';

> 다음의 경우엔 **Predicate Push**\가 수행되지 않는다.

Reviewer: ```suggestion
다음의 경우에는 **Predicate Push**\가 수행되지 않는다.
```

이를 통해 조건에 만족하는 더 적은 양의 데이터만 조회되도록 하여 전체 처리량을 줄일 수 있다.

예를 들어 **GROUP BY**\절을 포함한 아래 뷰를 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면
**GROUP BY** 해야 할 데이터양을 줄일 수 있다. 

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT * FROM athlete) a
        WHERE a.code = r.athlete_code
        AND a.nation_code = 'KOR';

위 쿼리에 정의한 뷰 내부에는 조건절이 없다. 만약 쿼리 변환이 작동하지 않는다면,
*emp* 테이블을 Full Scan 하고서 **GROUP BY** 이후에 *deptno = 30* 조건을
필터링했을 것이다. 

하지만 **Predicate Push**\를 통해서 다음과 같이 쿼리가 변환된다면, 더 적은 양의
데이터만 조회되도록 최적화할 수 있다.

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT * FROM athlete WHERE a.nation_code = 'KOR') a
        WHERE a.code = r.athlete_code
        AND a.nation_code = 'KOR';

다음의 경우엔 **Predicate Push**\가 수행되지 않는다.

    #. **CONNECT BY**\를 포함한 경우

    #. 집계함수나 분석함수를 사용하는 경우

    #. **ROWNUM, LIMIT** 또는 **GROUPBY_NUM (), INST_NUM (), ORDERBY_NUM ()**\ 을 사용하는 경우

    #. **Correlated Subquery**\ 를 사용하여 작성된 경우

    #. 조건절에 부질의가 사용된 경우

    #. 뷰가 메소드를 포함한 경우

또한 **OUTER JOIN**\을 수행할 때 다음의 경우 **Predicate Push**\가 수행되지 않는다.

    #. **OUTER JOIN**\의 **ON**\절에 조건절이 작성된 경우

    #. 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 **NULL** 변환 함수를 사용할 경우

>     #. 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 부질의를 사용하는 경우

Reviewer 2: It is not related to outer join

Author: I moved subquery example and RANDOM () example.

                9           60          600  'Left- 806'                     6
               10           60          600  'Left- 906'                     6

> .. _pred_push

Reviewer: ```suggestion
.. _pred-push
```

> .. _pred-push

Reviewer: ```suggestion
.. _pred-push:
```

Predicate Push
-----------------------

> **Predicate Push**\는 뷰를 조회하기 전에 먼저 조건절을 적용한다. 

Reviewer: I think the meaning of `뷰를 조회하기 전에 먼저 조건절을 적용한다.` is not clear.
This should mean using a predicate while querying the view, not before querying the view.

Author: I modified it to `**Predicate Push**\는 외부의 조건절을 뷰 안으로 밀어 넣는 최적화이다.`

> **Predicate Push**\는 뷰를 조회하기 전에 먼저 조건절을 적용한다. 

Reviewer 2: ```suggestion
**Predicate Push**\는 외부의 조건절을 뷰 안으로 밀어 넣는 최적화이다.
```

이를 통해 조건에 만족하는 더 적은 양의 데이터만 조회되도록 하여 전체 처리량을 줄일 수 있다.

> 예를 들어 **GROUP BY**\절을 포함한 아래 뷰를 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면

Reviewer: In changing the example, GROUP BY is not used.

Author: I replaced it to `예를 들어 아래 쿼리 처럼 *a.code = r.athlete_code*\절을 사용한 조인을 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면
조인해야 할 데이터양을 줄일 수 있다. `

**GROUP BY** 해야 할 데이터양을 줄일 수 있다. 

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT * FROM athlete) a
        WHERE a.code = r.athlete_code

>         AND a.nation_code = 'KOR';

Reviewer: There is no table referenced by `r` .

Author: I added table `record r`

위 쿼리에 정의한 뷰 내부에는 조건절이 없다. 만약 쿼리 변환이 작동하지 않는다면,

> *emp* 테이블을 Full Scan 하고서 **GROUP BY** 이후에 *deptno = 30* 조건을

Reviewer: The contents need to be changed to match the example.

Author: I modified it to `*athlete* 테이블을 Full Scan 하고서 **GROUP BY** 이후에 *a.nation_code = 'KOR'* 조건을
필터링했을 것이다. `

필터링했을 것이다. 

하지만 **Predicate Push**\를 통해서 다음과 같이 쿼리가 변환된다면, 더 적은 양의
데이터만 조회되도록 최적화할 수 있다.

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT * FROM athlete WHERE a.nation_code = 'KOR') a
        WHERE a.code = r.athlete_code

>         AND a.nation_code = 'KOR';

Reviewer: There is no table referenced by `r` .

Author: I added table record r

다음의 경우엔 **Predicate Push**\가 수행되지 않는다.

    #. **CONNECT BY**\를 포함한 경우

    #. 집계함수나 분석함수를 사용하는 경우

    #. **ROWNUM, LIMIT** 또는 **GROUPBY_NUM (), INST_NUM (), ORDERBY_NUM ()**\ 을 사용하는 경우

    #. **Correlated Subquery**\ 를 사용하여 작성된 경우

>     #. 조건절에 부질의가 사용된 경우

Reviewer: I think there needs to be consistency in terminology usage.
1. Correlated Subquery, Subquery
2. 상관 부질의(Correlated Subquery), 부질의(Subquery)

Using "Correlated Subquery" and "부질의" together seems terminologically inconsistent.

Reviewer 2: I think '부질의' is fine.

Reviewer 2: It's redundent '#. 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 부질의를 사용하는 경우'

    #. 뷰가 메소드를 포함한 경우

> 또한 **OUTER JOIN**\을 수행할 때 다음의 경우 **Predicate Push**가 수행되지 않는다.

Reviewer: ```suggestion
또한 **OUTER JOIN**\을 수행할 때 다음의 경우 **Predicate Push**\가 수행되지 않는다.
```

    #. **OUTER JOIN**\의 **ON**\절에 조건절이 작성된 경우

>     #. 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 NULL 변환 함수를 사용할 경우

Reviewer: ```suggestion
    #. 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 **NULL** 변환 함수를 사용할 경우
```

    #. 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 부질의를 사용하는 경우

    #. 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 **RANDOM (), DRANDOM (), SYS_GUID ()**\를 사용하는 경우

> 다음은 질의가 **OUTER JOIN**을 수행할 때 **ON**\절의 조건에 푸시될 조건절이 있는 예시이다.

Reviewer: ```suggestion
다음은 질의가 **OUTER JOIN**\을 수행할 때 **ON**\절의 조건에 푸시될 조건절이 있는 예시이다.
```

        SELECT a.name, r.score 

>         FROM (SELECT * FROM athlete) a

Reviewer: This example results in view merging before even considering predicate push.
Even if the `/*+ no_merge */` hint is used, views are merged.
So please consider adding the predicate below.

```suggestion
        FROM (SELECT * FROM athlete WHERE gender = 'M') a
```

>         SELECT a.name, r.score 
>         FROM (SELECT * FROM athlete WHERE gender = 'M') a
>                 LEFT OUTER JOIN record r ON a.code = r.athlete_code
>         AND a.nation_code = 'KOR';

Reviewer: ```suggestion
        SELECT a.name, r.score 
        FROM (SELECT * FROM athlete WHERE gender = 'M') a
            LEFT OUTER JOIN record r ON a.code = r.athlete_code AND a.nation_code = 'KOR';
```
AND is used with the ON clause.
Please use the same indentation as the ON clause. Or no line breaks required.

> 이 경우, a.nation_code = 'KOR'는 LEFT **OUTER JOIN** 수행 시 ON 절에 있는데, 이러한 형태로 ON 절의 조건절은 **Predicate Push** 대상이 아니다.

Reviewer: ```suggestion
이 경우, *a.nation_code = 'KOR'*\는 LEFT **OUTER JOIN** 수행 시 **ON** 절에 있는데, 이러한 형태로 **ON** 절의 조건절은 **Predicate Push** 대상이 아니다.
```

> 다음 질의는 **OUTER JOIN**\을 수행할 때 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 NULL 변환 함수를 사용하는 예시이다.

Reviewer: ```suggestion
다음 질의는 **OUTER JOIN**\을 수행할 때 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 **NULL** 변환 함수를 사용하는 예시이다.
```

.. code-block:: sql

        SELECT a.name, r.score 
        FROM athlete a

>         LEFT OUTER JOIN (SELECT * FROM record) r 

Reviewer: This example results in view merging before even considering predicate push.
Even if the /*+ no_merge */ hint is used, views are merged.
So please consider adding the predicate below.

```suggestion
        LEFT OUTER JOIN (SELECT * FROM record WHERE medal = 'G') r 
```

>         SELECT a.name, r.score 
>         FROM athlete a
>         LEFT OUTER JOIN (SELECT * FROM record) r 
>         ON a.code = r.athlete_code
>         WHERE NVL(r.score, '0') = '0';

Reviewer: It reduces readability.
I wish there was some formatting.

Below is an example.
```suggestion
        SELECT a.name, r.score 
        FROM athlete a
            LEFT OUTER JOIN (SELECT * FROM record) r ON a.code = r.athlete_code
        WHERE NVL(r.score, '0') = '0';
```

Author: I formatted OUTER JOIN like example.

> **OUTER JOIN**\을 수행할 때 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 NULL 변환 함수를 사용한 경우 **Predicate Push** 대상이 아니다.

Reviewer: ```suggestion
**OUTER JOIN**\을 수행할 때 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 **NULL** 변환 함수를 사용한 경우 **Predicate Push** 대상이 아니다.
```

                9           60          600  'Left- 806'                     6
               10           60          600  'Left- 906'                     6

.. _pred-push:

Predicate Push
-----------------------
**Predicate Push**\는 외부의 조건절을 뷰 안으로 밀어 넣는 최적화이다.

> 이를 통해 조건에 만족하는 더 적은 양의 데이터만 조회되도록 하여 전체 처리량을 줄일 수 있다.

Reviewer 4: ```suggestion
이를 통해 뷰 수행시 추가된 조건에 의해 더 적은 양의 데이터만 조회할 수 있어 전체 처리량을 줄일 수 있다.
```

> 예를 들어 아래 쿼리 처럼 *a.code = r.athlete_code*\의 조인조건으로 조인을 처리할 때, 쿼리 블록 밖에 있는 조건절을 쿼리 블록 안쪽에 밀어 넣을 수 있다면

Reviewer 4: ```suggestion
예를 들어 아래 질의 처럼 인라인 뷰와 테이블을 *a.code = r.athlete_code*\의 조인 조건으로 조인을 처리할 때, 조건절 중 인라인 뷰에 해당되는 조건절들을 인라인 뷰 안쪽에 밀어 넣을 수 있다면
```

조인해야 할 데이터양을 줄일 수 있다. 

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT name, nation_code, code, count(*) cnt FROM athlete GROUP BY name, nation_code) a, record r
        WHERE a.code = r.athlete_code
        AND a.nation_code = 'KOR';

> 위 쿼리에 정의한 뷰 내부에는 조건절이 없다. 만약 쿼리 변환이 작동하지 않는다면,

Reviewer 4: ```suggestion
위 질의 중 인라인 뷰 내부에는 조건절이 없다. 만약 쿼리 변환이 수행되지 않았다면,
```

> *athlete* 테이블을 Full Scan 하고서 조인을 수행한 이후에 *a.nation_code = 'KOR'* 조건을

Reviewer 4: ```suggestion
*athlete* 테이블을 full scan하여 결과와 record 테이블을 조인 수행후에 *a.nation_code = 'KOR'* 조건을
```

> *athlete* 테이블을 full scan하여 결과와 record 테이블을 조인 수행후에 *a.nation_code = 'KOR'* 조건을

Reviewer 4: ```suggestion
*athlete* 테이블을 full scan하여 생성된 임시 결과와 record 테이블을 조인 수행후에 *a.nation_code = 'KOR'* 조건을
```

필터링했을 것이다. 

> 하지만 **Predicate Push**\를 통해서 다음과 같이 쿼리가 변환된다면, 더 적은 양의

Reviewer 4: ```suggestion
하지만 **Predicate Push**\를 통해서 다음과 같이 질의가 변환된다면, 더 적은 양의
```

데이터만 조회되도록 최적화할 수 있다.

.. code-block:: sql

        SELECT a.name, r.score 
        FROM (SELECT name, nation_code, code, count(*) cnt FROM athlete WHERE nation_code = 'KOR' GROUP BY name, nation_code ) a, record r
        WHERE a.code = r.athlete_code;

다음의 경우에는 **Predicate Push**\가 수행되지 않는다.

>     #. 주질의에 **NO_PUSH_PRED** 힌트가 사용된 경우

Reviewer 4: ```suggestion
    #. 주 질의에 **NO_PUSH_PRED** 힌트가 사용된 경우
```

    #. **CONNECT BY**\를 포함한 경우

    #. 집계함수나 분석함수를 사용하는 경우

    #. **ROWNUM, LIMIT** 또는 **GROUPBY_NUM (), INST_NUM (), ORDERBY_NUM ()**\ 을 사용하는 경우

    #. **Correlated Subquery**\ 를 사용하여 작성된 경우

    #. 조건절에 부질의가 사용된 경우

    #. 뷰가 메소드를 포함한 경우

    #. 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 **RANDOM (), DRANDOM (), SYS_GUID ()**\를 사용하는 경우

    #. **OUTER JOIN**\을 수행할 때 푸시될 조건절이나 뷰 내부의 **Predicate Push** 대상에 다음을 사용하는 경우:
            * **ON**\절에 조건절이 작성된 경우
            * **NULL** 변환 함수

> 다음은 주질의에 NO_PUSH_PRED 힌트가 사용된 예시이다.

Reviewer 4: ```suggestion
다음은 주 질의에 NO_PUSH_PRED 힌트가 사용된 예시이다.
```

.. code-block:: sql

        SELECT /*+ NO_PUSH_PRED*/ a.name, r.score 
        FROM (SELECT name, nation_code, code, count(*) cnt FROM athlete GROUP BY name, nation_code) a, record r
        WHERE a.code = r.athlete_code
        AND a.nation_code = 'KOR';

> 주질의에 NO_PUSH_PRED 힌트가 사용된 경우, **Predicate Push**\가 수행되지 않는다.

Reviewer 4: ```suggestion
주 질의에 NO_PUSH_PRED 힌트가 사용된 경우, **Predicate Push**\가 수행되지 않는다.
```
